core/src/main/java/org/apache/calcite/runtime/ResultSetEnumerable.java

```diff
@@ -146,19 +174,55 @@ private ResultSetEnumerable(
     } catch (SQLException e) {
       throw new RuntimeException("while executing SQL [" + sql + "]", e);
     } finally {
-      if (statement != null) {
-        try {
-          statement.close();
-        } catch (SQLException e) {
-          // ignore
-        }
+      closeIfPossible(connection, statement);
+    }
+  }
+
+  private Enumerator<T> enumeratorBasedOnPeparedStatement() {
+    Connection connection = null;
+    PreparedStatement statement = null;
+    try {
+      connection = dataSource.getConnection();
+      statement = connection.prepareStatement(sql);
+      setTimeoutIfPossible(statement);
+      consumer.accept(statement);
+      if (statement.execute()) {
+        final ResultSet resultSet = statement.getResultSet();
+        statement = null;
+        connection = null;
+        return new ResultSetEnumerator<>(resultSet, rowBuilderFactory);
+      } else {
+        Integer updateCount = statement.getUpdateCount();
+        return Linq4j.singletonEnumerator((T) updateCount);
       }
-      if (connection != null) {
-        try {
-          connection.close();
-        } catch (SQLException e) {
-          // ignore
-        }
+    } catch (SQLException e) {
+      throw new RuntimeException("while executing SQL [" + sql + "]", e);
```

Comment:
I wonder if something like `src/main/java/org/apache/calcite/runtime/CalciteResource.java` should be used instead